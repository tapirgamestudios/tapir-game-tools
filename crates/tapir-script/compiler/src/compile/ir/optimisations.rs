use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
