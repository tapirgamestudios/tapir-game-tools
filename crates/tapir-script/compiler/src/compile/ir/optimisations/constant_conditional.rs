use std::collections::HashMap;

use crate::{
    ast::SymbolId,
    compile::ir::{
        BlockExitInstr, Constant, TapIrFunction, TapIrFunctionBlockIter, TapIrInstr,
        optimisations::OptimisationResult,
    },
};

pub fn remove_constant_conditionals(f: &mut TapIrFunction) -> OptimisationResult {
    let mut did_something = OptimisationResult::DidNothing;

    let mut constants: HashMap<SymbolId, bool> = HashMap::new();

    let mut reverse_post_order = TapIrFunctionBlockIter::new_reverse_post_order(f);
    while let Some(block) = reverse_post_order.next_mut(f) {
        for instr in block.instrs() {
            if let TapIrInstr::Constant(target, Constant::Bool(value)) = instr.instr {
                constants.insert(target, value);
            }
        }

        if let BlockExitInstr::ConditionalJump {
            test,
            if_true,
            if_false,
        } = block.block_exit()
            && let Some(value) = constants.get(test)
        {
            let target_block = if *value { *if_true } else { *if_false };

            did_something = OptimisationResult::DidSomething;
            *block.block_exit_mut() = BlockExitInstr::JumpToBlock(target_block);
        }
    }

    did_something
}
