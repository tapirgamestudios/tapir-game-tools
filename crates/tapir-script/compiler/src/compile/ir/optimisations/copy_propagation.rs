use std::collections::HashMap;

use crate::{
    ast::SymbolId,
    compile::ir::{
        TapIrFunction, TapIrFunctionBlockIter, TapIrInstr, optimisations::OptimisationResult,
    },
};

/// Finds all the moves copies the target of the move wherever possible.
///
/// A subsequent optimisation pass removes the dead stores that result from this pass.
pub fn copy_propagation(f: &mut TapIrFunction) -> OptimisationResult {
    let mut renames: HashMap<SymbolId, SymbolId> = HashMap::new();

    let mut did_something = OptimisationResult::DidNothing;

    macro_rules! replace_source {
        ($source:ident) => {
            if let Some(new_source) = renames.get($source) {
                *$source = *new_source;
                did_something = OptimisationResult::DidSomething;
            }
        };
    }

    // reverse post order is the best order to do this, but this will require multiple passes anyway
    let mut dfs = TapIrFunctionBlockIter::new_reverse_post_order(f);
    while let Some(block) = dfs.next_mut(f) {
        for phi in block.block_entry_mut() {
            for (_, source) in &mut phi.sources {
                replace_source!(source);
            }
        }

        for instr in block.instrs_mut() {
            for source in instr.sources_mut() {
                replace_source!(source);
            }

            if let TapIrInstr::Move { source, target } = instr.instr {
                if source != target && renames.insert(target, source).is_some() {
                    panic!(
                        "Should be in SSA so there should only ever be one definition of source"
                    );
                }
            }
        }

        for source in block.block_exit_mut().sources_mut() {
            replace_source!(source);
        }
    }

    did_something
}
