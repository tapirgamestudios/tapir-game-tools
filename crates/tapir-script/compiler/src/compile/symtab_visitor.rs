use std::{borrow::Cow, collections::HashMap};

use crate::{
    ast::{
        Expression, ExpressionKind, Function, FunctionId, MaybeResolved, Statement, StatementKind,
        SymbolId,
    },
    reporting::{CompilerErrorKind, Diagnostics},
    tokens::Span,
};

use super::{CompileSettings, Property};

pub struct SymTabVisitor<'input> {
    symtab: SymTab<'input>,

    symbol_names: NameTable<'input>,
    function_names: HashMap<&'input str, FunctionId>,
}

impl<'input> SymTabVisitor<'input> {
    pub fn new(
        settings: &CompileSettings,
        functions: &mut [Function<'input>],
        diagnostics: &mut Diagnostics,
    ) -> Self {
        let mut function_declarations = HashMap::new();
        let mut function_names = vec![];

        for (i, function) in functions.iter_mut().enumerate() {
            function.meta.set(FunctionId(i));
            function_names.push(function.name);

            if let Some(other_span) = function_declarations.insert(function.name, function.span) {
                diagnostics.add_message(
                    CompilerErrorKind::FunctionAlreadyDeclared {
                        function_name: function.name.to_string(),
                        old_function_declaration: other_span,
                        new_function_declaration: function.span,
                    }
                    .into_message(function.span),
                );
            }
        }

        Self {
            symtab: SymTab::new(settings, function_names),
            symbol_names: NameTable::new(settings),
            function_names: functions
                .iter()
                .map(|f| (f.name, *f.meta.get::<FunctionId>().unwrap()))
                .collect(),
        }
    }

    pub fn get_symtab(&self) -> &SymTab<'input> {
        &self.symtab
    }

    pub fn into_symtab(self) -> SymTab<'input> {
        self.symtab
    }

    pub fn visit_function(
        &mut self,
        function: &mut Function<'input>,
        diagnostics: &mut Diagnostics,
    ) {
        self.symbol_names.push_scope();

        for argument in &mut function.arguments {
            let MaybeResolved::Unresolved(name) = argument.name else {
                panic!("Should not have resolved arguments yet");
            };

            let symbol_id = self.symtab.new_symbol(name, argument.span);
            self.symbol_names.insert(name, symbol_id);

            argument.name = MaybeResolved::Resolved(symbol_id);
        }

        self.visit_block(&mut function.statements, diagnostics);

        self.symbol_names.pop_scope();
    }

    fn visit_block(&mut self, ast: &mut [Statement<'input>], diagnostics: &mut Diagnostics) {
        self.symbol_names.push_scope();

        for statement in ast {
            match &mut statement.kind {
                StatementKind::VariableDeclaration { ident, value } => {
                    self.visit_expr(value, diagnostics);

                    let symbol_id = self.symtab.new_symbol(ident, statement.span);
                    self.symbol_names.insert(ident, symbol_id);

                    statement.meta.set(symbol_id);
                }
                StatementKind::Assignment { ident, value } => {
                    self.visit_expr(value, diagnostics);

                    if let Some(symbol_id) = self.symbol_names.get(ident) {
                        statement.meta.set(symbol_id);
                    } else {
                        diagnostics.add_message(
                            CompilerErrorKind::UnknownVariable(ident.to_string())
                                .into_message(statement.span),
                        );
                    }
                }
                StatementKind::If {
                    condition,
                    true_block,
                    false_block,
                } => {
                    self.visit_expr(condition, diagnostics);

                    self.visit_block(true_block, diagnostics);
                    self.visit_block(false_block, diagnostics);
                }
                StatementKind::Error
                | StatementKind::Wait
                | StatementKind::Nop
                | StatementKind::Continue
                | StatementKind::Break => {}
                StatementKind::Return { values } => {
                    for expr in values {
                        self.visit_expr(expr, diagnostics);
                    }
                }
                StatementKind::Block { block } => {
                    self.visit_block(block, diagnostics);
                }
                StatementKind::Call { arguments, name }
                | StatementKind::Spawn { arguments, name } => {
                    if let Some(function) = self.function_names.get(name) {
                        statement.meta.set(*function);
                    } else {
                        diagnostics.add_message(
                            CompilerErrorKind::UnknownFunction {
                                name: name.to_string(),
                            }
                            .into_message(statement.span),
                        );
                    }

                    for argument in arguments {
                        self.visit_expr(argument, diagnostics);
                    }
                }
                StatementKind::Trigger { arguments, .. } => {
                    for argument in arguments {
                        self.visit_expr(argument, diagnostics);
                    }
                }
                StatementKind::Loop { block } => {
                    self.visit_block(block, diagnostics);
                }
            };
        }

        self.symbol_names.pop_scope();
    }

    fn visit_expr(&self, expr: &mut Expression<'_>, diagnostics: &mut Diagnostics) {
        match &mut expr.kind {
            ExpressionKind::Variable(ident) => {
                if let Some(symbol_id) = self.symbol_names.get(ident) {
                    expr.meta.set(symbol_id);
                } else {
                    diagnostics.add_message(
                        CompilerErrorKind::UnknownVariable(ident.to_string())
                            .into_message(expr.span),
                    );
                }
            }
            ExpressionKind::BinaryOperation { lhs, rhs, .. } => {
                self.visit_expr(lhs, diagnostics);
                self.visit_expr(rhs, diagnostics);
            }
            ExpressionKind::Call { arguments, name } => {
                if let Some(function) = self.function_names.get(name) {
                    expr.meta.set(*function);
                } else {
                    diagnostics.add_message(
                        CompilerErrorKind::UnknownFunction {
                            name: name.to_string(),
                        }
                        .into_message(expr.span),
                    );
                }

                for argument in arguments {
                    self.visit_expr(argument, diagnostics);
                }
            }
            ExpressionKind::Integer(_)
            | ExpressionKind::Fix(_)
            | ExpressionKind::Error
            | ExpressionKind::Nop
            | ExpressionKind::Bool(_) => {}
        }
    }
}

struct NameTable<'input> {
    names: Vec<HashMap<Cow<'input, str>, SymbolId>>,
}

impl<'input> NameTable<'input> {
    pub fn new(settings: &CompileSettings) -> Self {
        let property_symbols = settings
            .properties
            .iter()
            .enumerate()
            .map(|(i, prop)| (Cow::Owned(prop.name.clone()), SymbolId(i)))
            .collect();

        Self {
            names: vec![property_symbols],
        }
    }

    pub fn insert(&mut self, name: &'input str, id: SymbolId) {
        self.names
            .last_mut()
            .unwrap()
            .insert(Cow::Borrowed(name), id);
    }

    pub fn get(&self, name: &str) -> Option<SymbolId> {
        for nametab in self.names.iter().rev() {
            if let Some(id) = nametab.get(name) {
                return Some(*id);
            }
        }

        None
    }

    pub fn push_scope(&mut self) {
        self.names.push(HashMap::new())
    }

    pub fn pop_scope(&mut self) {
        self.names.pop();
        assert!(!self.names.is_empty());
    }
}

pub struct SymTab<'input> {
    properties: Vec<Property>,

    symbol_names: Vec<(Cow<'input, str>, Option<Span>)>,

    // function names are only used in tests
    #[cfg_attr(not(test), expect(dead_code))]
    function_names: Vec<&'input str>,
}

impl<'input> SymTab<'input> {
    fn new(settings: &CompileSettings, function_names: Vec<&'input str>) -> Self {
        let properties = settings.properties.clone();
        let symbol_names = properties
            .iter()
            .map(|prop| (Cow::Owned(prop.name.clone()), None))
            .collect();

        Self {
            properties,
            symbol_names,
            function_names,
        }
    }

    fn new_symbol(&mut self, ident: &'input str, span: Span) -> SymbolId {
        self.symbol_names.push((Cow::Borrowed(ident), Some(span)));
        SymbolId(self.symbol_names.len() - 1)
    }

    pub(crate) fn new_rename(&mut self, symbol_id: SymbolId) -> SymbolId {
        self.symbol_names
            .push(self.symbol_names[symbol_id.0].clone());
        SymbolId(self.symbol_names.len() - 1)
    }

    pub(crate) fn new_temporary(&mut self) -> SymbolId {
        let id = self.symbol_names.len();
        self.symbol_names.push((Cow::Borrowed(""), None));
        SymbolId(id)
    }

    pub(crate) fn name_for_symbol(&self, symbol_id: SymbolId) -> Cow<'input, str> {
        let name = self.symbol_names[symbol_id.0].0.clone();
        if name.is_empty() {
            Cow::Owned(format!("temp.{}", symbol_id.0))
        } else {
            name
        }
    }

    #[cfg(test)]
    pub(crate) fn debug_name_for_symbol(&self, symbol_id: SymbolId) -> String {
        let name = self.symbol_names[symbol_id.0].0.clone();
        if name.is_empty() {
            format!("temp.{}", symbol_id.0)
        } else {
            format!("{name}.{}", symbol_id.0)
        }
    }

    #[cfg(test)]
    pub(crate) fn name_for_function(&self, function_id: FunctionId) -> &'input str {
        self.function_names[function_id.0]
    }

    pub(crate) fn span_for_symbol(&self, symbol_id: SymbolId) -> Span {
        self.symbol_names[symbol_id.0]
            .1
            .expect("Symbol should have a span")
    }

    #[cfg(test)]
    pub fn all_symbols(&self) -> impl Iterator<Item = (&'_ str, SymbolId)> + '_ {
        self.symbol_names
            .iter()
            .enumerate()
            .map(|(i, (name, _span))| (name.as_ref(), SymbolId(i)))
    }

    pub fn get_property(&self, symbol_id: SymbolId) -> Option<&Property> {
        self.properties.get(symbol_id.0)
    }
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_ron_snapshot, assert_snapshot, glob};

    use crate::{grammar, lexer::Lexer, tokens::FileId, types::Type};

    use super::*;

    #[test]
    fn symtab_success_snapshot_tests() {
        glob!("snapshot_tests", "symtab_visitor/*_success.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let parser = grammar::ScriptParser::new();
            let file_id = FileId::new(0);

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

            let mut visitor = SymTabVisitor::new(
                &CompileSettings {
                    properties: vec![Property {
                        ty: Type::Int,
                        index: 0,
                        name: "int_prop".to_string(),
                    }],
                    enable_optimisations: false,
                },
                &mut script.functions,
                &mut diagnostics,
            );

            for function in &mut script.functions {
                visitor.visit_function(function, &mut diagnostics);
            }

            assert_ron_snapshot!(script, {
                ".**.span" => "[span]",
            });
        });
    }

    #[test]
    fn symtab_fail_snapshot_tests() {
        glob!("snapshot_tests", "symtab_visitor/*_fail.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let file_id = FileId::new(0);
            let lexer = Lexer::new(&input, file_id);
            let parser = grammar::ScriptParser::new();

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser
                .parse(FileId::new(0), &mut diagnostics, lexer)
                .unwrap();

            let mut visitor = SymTabVisitor::new(
                &CompileSettings {
                    properties: vec![Property {
                        ty: Type::Int,
                        index: 0,
                        name: "int_prop".to_string(),
                    }],
                    enable_optimisations: false,
                },
                &mut script.functions,
                &mut diagnostics,
            );

            for function in &mut script.functions {
                visitor.visit_function(function, &mut diagnostics);
            }

            assert_snapshot!(diagnostics.pretty_string(false));
        });
    }
}
