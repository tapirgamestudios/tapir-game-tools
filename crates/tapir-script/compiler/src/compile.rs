use std::{collections::HashMap, path::Path};

use bytecode::{Type1, Type3};
use optimisations::UnusedFunction;
use symtab_visitor::SymTabVisitor;
use type_visitor::{TypeTable, TypeVisitor};

use crate::{
    ast::{BinaryOperator, FunctionId, SymbolId},
    compile::ir::{create_ir, BlockId, TapIrFunction, TapIrInstr},
    grammar,
    lexer::Lexer,
    reporting::Diagnostics,
    tokens::FileId,
    types::Type,
    EventHandler, Trigger,
};

#[cfg(test)]
mod disassemble;
mod ir;
mod loop_visitor;
mod optimisations;
mod symtab_visitor;
mod type_visitor;

use bytecode::Opcode;

#[derive(Clone, Debug)]
pub struct Property {
    pub ty: Type,
    pub index: usize,
    pub name: String,
}

pub struct CompileSettings {
    pub properties: Vec<Property>,
    pub enable_optimisations: bool,
}

impl CompileSettings {
    pub(crate) fn is_property(&self, symbol_id: SymbolId) -> bool {
        symbol_id.0 < self.properties.len()
    }

    pub(crate) fn property_symbols(&self) -> impl Iterator<Item = SymbolId> {
        (0..self.properties.len()).map(SymbolId)
    }
}

pub fn compile(
    filename: impl AsRef<Path>,
    input: &str,
    settings: &CompileSettings,
) -> Result<Bytecode, Diagnostics> {
    let file_id = FileId::new(0);

    let mut diagnostics = Diagnostics::new(file_id, filename, input);

    let lexer = Lexer::new(input, file_id);
    let parser = grammar::ScriptParser::new();

    let mut ast = match parser.parse(file_id, &mut diagnostics, lexer) {
        Ok(ast) => ast,
        Err(e) => {
            diagnostics.add_lalrpop(e, file_id);
            return Err(diagnostics);
        }
    };

    let mut sym_tab_visitor = SymTabVisitor::new(settings, &mut ast.functions, &mut diagnostics);
    let mut type_visitor = TypeVisitor::new(settings, &ast.functions);

    for function in &mut ast.functions {
        sym_tab_visitor.visit_function(function, &mut diagnostics);
        loop_visitor::visit_loop_check(function, &mut diagnostics);

        type_visitor.visit_function(function, sym_tab_visitor.get_symtab(), &mut diagnostics);
    }

    optimisations::optimise(&mut ast.functions, settings, &mut diagnostics);

    let type_table = type_visitor.into_type_table(sym_tab_visitor.get_symtab(), &mut diagnostics);

    if diagnostics.has_any() {
        return Err(diagnostics);
    }

    let mut compiler = Compiler::new(&type_table);
    let mut symtab = sym_tab_visitor.into_symtab();

    let ir_functions = ast
        .functions
        .iter()
        .filter(|f| !f.meta.has::<UnusedFunction>())
        .map(|f| create_ir(f, &mut symtab))
        .collect::<Vec<_>>();

    for function in ir_functions {
        compiler.compile_function(&function);
    }

    Ok(compiler.finalise())
}

struct Compiler {
    stack: Vec<Option<SymbolId>>,

    function_locations: HashMap<FunctionId, Label>,
    jumps: Vec<(FunctionId, Jump)>,

    bytecode: Bytecode,
}

impl Compiler {
    pub fn new(type_table: &TypeTable<'_>) -> Self {
        Self {
            stack: vec![],

            jumps: vec![],
            function_locations: HashMap::from([(FunctionId(0), Label(0))]),
            bytecode: Bytecode::new(type_table.triggers()),
        }
    }

    pub fn compile_function(&mut self, function: &TapIrFunction) {
        let function_id = function.id;

        if function_id != FunctionId(0) {
            // the stack will be arguments, then the return pointer. However, if we're at toplevel, then
            // the stack will be empty to start with
            for argument in &function.arguments {
                self.stack.push(Some(*argument));
            }

            self.stack.push(None);
        }

        self.function_locations
            .insert(function_id, self.bytecode.new_label());

        if let Some(event_handler) = &function.modifiers.event_handler {
            self.bytecode.event_handlers.push(EventHandler {
                name: event_handler.name.clone(),
                bytecode_offset: self.bytecode.offset(),
                arguments: event_handler.arg_names.clone(),
            });
        }

        self.compile_blocks(function);

        // if there is no return value, then no return is required to be compiled so we should add one
        if function.return_types.is_empty() {
            self.bytecode.ret();
        }
    }

    fn compile_blocks(&mut self, function: &TapIrFunction) {
        // Where all the blocks start
        let mut block_entrypoints: HashMap<BlockId, Label> = HashMap::new();
        // Where all the jumps are and to which block they should go. For non-local jumps,
        // these are filled in later as function calls.
        let mut jumps: Vec<(BlockId, Jump)> = vec![];

        let mut symbols: Vec<_> = function.symbols().iter().copied().collect();
        symbols.sort_by_key(|sym| sym.0);
        assert!(
            symbols.len() < 200,
            "Need fewer than 200 symbols in a function, but got {}",
            symbols.len()
        );
        let var_locations = symbols
            .iter()
            .enumerate()
            .map(|(i, sym)| {
                if let Some(i) = function.arguments.iter().position(|arg| arg == sym) {
                    (*sym, i as u8 + 1)
                } else {
                    (*sym, i as u8 + function.arguments.len() as u8 + 1)
                }
            })
            .collect::<HashMap<_, _>>();

        let v = move |s: &SymbolId| *var_locations.get(s).unwrap();

        let first_argument = symbols.len() as u8 + function.arguments.len() as u8 + 1;
        let put_args = |bytecode: &mut Bytecode, args: &[SymbolId]| {
            for (i, arg) in args.iter().enumerate() {
                bytecode.mov(i as u8 + first_argument + 1, v(arg));
            }
        };

        for block in &function.blocks {
            let entry_point = self.bytecode.new_label();
            block_entrypoints.insert(block.id, entry_point);

            for instr in &block.instrs {
                match &instr.instr {
                    TapIrInstr::Constant(target, constant) => {
                        let constant = match constant {
                            ir::Constant::Int(i) => *i as u32,
                            ir::Constant::Fix(num) => num.to_raw() as u32,
                            ir::Constant::Bool(b) => *b as u32,
                        };

                        self.bytecode.constant(v(target), constant);
                    }
                    TapIrInstr::Move { target, source } => self.bytecode.mov(v(target), v(source)),
                    TapIrInstr::BinOp {
                        target,
                        lhs,
                        op,
                        rhs,
                    } => {
                        self.bytecode.binop(v(target), v(lhs), *op, v(rhs));
                    }
                    TapIrInstr::Wait => self.bytecode.wait(),
                    TapIrInstr::Call { target, f, args } => {
                        put_args(&mut self.bytecode, args);
                        self.bytecode.call(first_argument);
                        self.jumps.push((*f, self.bytecode.new_jump()));

                        for (i, target) in target.iter().enumerate() {
                            self.bytecode.mov(v(target), i as u8 + first_argument + 1);
                        }
                    }
                    TapIrInstr::Spawn { f, args } => {
                        put_args(&mut self.bytecode, args);
                        self.bytecode.spawn(first_argument, args.len() as u8);
                        self.jumps.push((*f, self.bytecode.new_jump()));
                    }
                    TapIrInstr::Trigger { f, args } => {
                        put_args(&mut self.bytecode, args);
                        self.bytecode.trigger(f.0 as u8, first_argument);
                    }
                    TapIrInstr::GetProp { target, prop_index } => {
                        self.bytecode.get_prop(v(target), *prop_index as u8);
                    }
                    TapIrInstr::StoreProp { prop_index, value } => {
                        self.bytecode.set_prop(v(value), *prop_index as u8);
                    }
                }
            }

            match &block.block_exit {
                ir::BlockExitInstr::JumpToBlock(block_id) => {
                    let jump = self.bytecode.new_jump();
                    jumps.push((*block_id, jump));
                }
                ir::BlockExitInstr::ConditionalJump {
                    test,
                    if_true,
                    if_false,
                } => {
                    self.bytecode.jump_if(v(test));
                    let true_jump = self.bytecode.new_jump();
                    let false_jump = self.bytecode.new_jump();

                    jumps.extend([(*if_true, true_jump), (*if_false, false_jump)]);
                }
                ir::BlockExitInstr::Return(symbol_ids) => {
                    for (i, symbol) in symbol_ids.iter().enumerate() {
                        self.bytecode.mov(i as u8 + 1, v(symbol));
                    }

                    self.bytecode.ret();
                }
            }
        }

        for (block_id, jump) in jumps {
            let label = block_entrypoints
                .get(&block_id)
                .expect("Should jump to a valid block");

            self.bytecode.patch_jump(jump, *label);
        }
    }

    fn finalise(mut self) -> Bytecode {
        for (function_id, jump) in self.jumps {
            let label = self
                .function_locations
                .get(&function_id)
                .expect("Should have defined a function");
            self.bytecode.patch_jump(jump, *label);
        }

        self.bytecode
    }
}

pub struct Bytecode {
    pub data: Vec<u32>,

    pub event_handlers: Vec<EventHandler>,
    pub triggers: Box<[Trigger]>,
}

impl Bytecode {
    fn new(triggers: Box<[Trigger]>) -> Self {
        Self {
            data: vec![],

            event_handlers: vec![],
            triggers,
        }
    }

    fn offset(&self) -> usize {
        self.data.len()
    }

    fn new_label(&self) -> Label {
        Label(self.data.len())
    }

    fn new_jump(&mut self) -> Jump {
        self.data.push(Type3::invalid_jump().encode());
        Jump(self.data.len() - 1)
    }

    fn patch_jump(&mut self, jump: Jump, label: Label) {
        assert_eq!(
            bytecode::opcode(self.data[jump.0]),
            Some(Opcode::Jump),
            "Trying to patch a non-jump instruction"
        );

        self.data[jump.0] =
            Type3::jump(label.0.try_into().expect("Offset bigger than allowed")).encode();
    }
}

impl Bytecode {
    fn ret(&mut self) {
        self.data.push(Type1::ret().encode());
    }

    fn mov(&mut self, target: u8, source: u8) {
        self.data.push(Type1::mov(target, source).encode());
    }

    fn wait(&mut self) {
        self.data.push(Type1::wait().encode());
    }

    fn constant(&mut self, target: u8, value: u32) {
        self.data.push(Type1::constant(target).encode());
        self.data.push(value);
    }

    fn call(&mut self, first_arg: u8) {
        self.data.push(Type1::call(first_arg).encode());
    }

    fn spawn(&mut self, first_arg: u8, num_args: u8) {
        self.data.push(Type1::spawn(first_arg, num_args).encode());
    }

    fn trigger(&mut self, id: u8, first_arg: u8) {
        self.data.push(Type1::trigger(id, first_arg).encode());
    }

    fn jump_if(&mut self, target: u8) {
        self.data.push(Type1::jump_if(target).encode());
    }

    fn get_prop(&mut self, target: u8, prop_index: u8) {
        self.data.push(Type1::get_prop(target, prop_index).encode());
    }

    fn set_prop(&mut self, target: u8, prop_index: u8) {
        self.data.push(Type1::set_prop(target, prop_index).encode());
    }

    fn binop(&mut self, target: u8, lhs: u8, binop: BinaryOperator, rhs: u8) {
        let opcode = match binop {
            BinaryOperator::Add => Opcode::Add,
            BinaryOperator::Sub => Opcode::Sub,
            BinaryOperator::Mul => Opcode::Mul,
            BinaryOperator::Div => unreachable!("Shouldn't be compiling div binops"),
            BinaryOperator::Mod => unreachable!("Shouldn't be compiling mod binops"),
            BinaryOperator::RealDiv => Opcode::RealDiv,
            BinaryOperator::RealMod => Opcode::RealMod,
            BinaryOperator::FixMul => Opcode::FixMul,
            BinaryOperator::FixDiv => Opcode::FixDiv,
            BinaryOperator::EqEq => Opcode::EqEq,
            BinaryOperator::NeEq => Opcode::NeEq,
            BinaryOperator::Gt => Opcode::Gt,
            BinaryOperator::GtEq => Opcode::GtEq,
            BinaryOperator::Lt => Opcode::Lt,
            BinaryOperator::LtEq => Opcode::LtEq,
            BinaryOperator::Then => unreachable!("Shouldn't be compiling then binops"),
        };

        self.data
            .push(Type1::binop(opcode, target, lhs, rhs).encode());
    }
}

// The internal usize here is the index in the bytecode where this label sits
#[derive(Clone, Copy)]
struct Label(usize);

// The internal usize here is the index where the jump instruction is in the bytecode
#[derive(Clone, Copy)]
struct Jump(usize);

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_snapshot, glob};

    use super::*;

    #[test]
    fn compiler_snapshot_tests() {
        glob!("snapshot_tests", "compiler/*.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let compiler_settings = CompileSettings {
                properties: vec![Property {
                    ty: Type::Int,
                    index: 0,
                    name: "int_prop".to_string(),
                }],
                enable_optimisations: false,
            };

            let bytecode = compile(path, &input, &compiler_settings).unwrap();

            let mut decompiled = String::new();
            disassemble::disassemble(&bytecode.data, &mut decompiled).unwrap();

            assert_snapshot!(decompiled);
        });
    }
}
