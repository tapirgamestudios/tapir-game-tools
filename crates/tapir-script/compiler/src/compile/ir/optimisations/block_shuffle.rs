use petgraph::{prelude::DiGraphMap, visit::IntoNeighbors};

use crate::compile::ir::{
    BlockExitInstr, TapIrFunction, TapIrFunctionBlockIter, optimisations::OptimisationResult,
};

pub fn simplify_blocks(f: &mut TapIrFunction) -> OptimisationResult {
    let mut did_something = OptimisationResult::DidNothing;

    // we need to know the predecessors of every block unfortunately
    let mut full_graph = DiGraphMap::new();
    for block in f.blocks() {
        for neighbour in f.neighbors(block.id()) {
            full_graph.add_edge(block.id(), neighbour, ());
        }
    }

    let mut dfs = TapIrFunctionBlockIter::new_dfs(f);
    while let Some(block_id) = dfs.next_id(f) {
        let block = f.block(block_id).unwrap();
        let BlockExitInstr::JumpToBlock(next_block_id) = block.block_exit() else {
            // this block doesn't have an unconditional jump in it
            continue;
        };

        let next_block_id = *next_block_id;

        if next_block_id == block_id {
            // this block references itself, so we can't do any shuffling
            continue;
        }

        if full_graph
            .edges_directed(next_block_id, petgraph::Direction::Incoming)
            .count()
            != 1
        {
            // we can't just copy the content of the next block into this one
            continue;
        }

        // copy the content of the next block into this one, since there isn't another block jumping to it
        let [this_block, next_block] = f.disjoint_blocks_mut([&block_id, &next_block_id]);

        if next_block.instrs.is_empty() {
            // nothing to do here since there aren't any instructions in next
            continue;
        }

        if !next_block.block_entry().is_empty() {
            // phis make this a little complex, so don't bother
            continue;
        }

        this_block.instrs.append(&mut next_block.instrs);

        did_something = OptimisationResult::DidSomething;
    }

    did_something
}
