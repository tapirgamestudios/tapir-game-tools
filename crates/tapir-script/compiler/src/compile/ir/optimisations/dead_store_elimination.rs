use std::collections::HashSet;

use crate::compile::ir::{
    TapIrFunction, TapIrFunctionBlockIter, optimisations::OptimisationResult,
};

/// Removes every instruction which stores but doesn't ultimately use the result
pub fn remove_dead_stores(f: &mut TapIrFunction) -> OptimisationResult {
    let used_symbols = f.blocks().flat_map(|b| b.sources()).collect::<HashSet<_>>();

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = TapIrFunctionBlockIter::new_dfs(f);
    while let Some(block) = dfs.next_mut(f) {
        block.block_entry_retain_mut(|phi| {
            if !used_symbols.contains(&phi.target) {
                did_something = OptimisationResult::DidSomething;
                return false;
            }

            true
        });

        block.instrs_retain(|instr| {
            if instr.could_have_side_effects() {
                return true;
            }

            if instr.targets().any(|target| used_symbols.contains(&target)) {
                return true;
            }

            did_something = OptimisationResult::DidSomething;
            false
        });
    }

    did_something
}
