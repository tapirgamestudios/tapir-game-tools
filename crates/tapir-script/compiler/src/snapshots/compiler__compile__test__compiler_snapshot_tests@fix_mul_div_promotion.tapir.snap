---
source: crates/tapir-script/compiler/src/compile.rs
expression: decompiled
input_file: crates/tapir-script/compiler/src/snapshot_tests/compiler/fix_mul_div_promotion.tapir
---
00000000: loadk 1, 128
00000002: mov 3, 1
00000003: loadk 4, 768
00000005: fixmul 2, 3, 4
00000006: ret
00000007: ret
